packages/engine/Source/Shaders/Voxels/VoxelFS.glsl

```diff
@@ -23,30 +23,41 @@ uniform float u_stepSize;
     uniform vec4 u_pickColor;
 #endif
 
-#if defined(JITTER)
-float hash(vec2 p)
-{
-    vec3 p3 = fract(vec3(p.xyx) * 50.0); // magic number = hashscale
-    p3 += dot(p3, p3.yzx + 19.19);
-    return fract((p3.x + p3.y) * p3.z);
+vec3 getSampleSize(in int level) {
+    vec3 sampleCount = exp2(float(level)) * vec3(u_dimensions);
+    vec3 sampleSizeUv = 1.0 / sampleCount;
+    return scaleShapeUvToShapeSpace(sampleSizeUv);
 }
-#endif
 
-vec4 getStepSize(in SampleData sampleData, in Ray viewRay, in RayShapeIntersection shapeIntersection, in float currentT) {
-#if defined(SHAPE_BOX)
-    Box voxelBox = constructVoxelBox(sampleData.tileCoords, sampleData.tileUv);
-    RayShapeIntersection voxelIntersection = intersectBox(viewRay, voxelBox);
-    vec4 entry = intersectionMax(shapeIntersection.entry, voxelIntersection.entry);
-    float exit = min(voxelIntersection.exit.w, shapeIntersection.exit.w);
-    float dt = (exit - entry.w) * RAY_SCALE;
-    return vec4(normalize(entry.xyz), dt);
-#else
-    float dimAtLevel = pow(2.0, float(sampleData.tileCoords.w));
-    float shapeEntryT = shapeIntersection.entry.w * RAY_SCALE;
-    float constantStep = u_stepSize / dimAtLevel;
-    vec3 normal = (currentT < shapeEntryT + constantStep) ? shapeIntersection.entry.xyz : viewRay.dir;
-    return vec4(normalize(normal), constantStep);
-#endif
+vec4 getStepSize(in SampleData sampleData, in Ray viewRay, in RayShapeIntersection shapeIntersection, in mat3 jacobianT, in float currentT) {
+    // The Jacobian is computed in a space where the ellipsoid spans [-1, 1].
+    // But the ray is marched in a space where the ellipsoid fills [0, 1].
+    // So we need to scale the Jacobian by 2.
+    vec3 gradient = 2.0 * viewRay.rawDir * jacobianT;
+
+    vec3 sampleSizeAlongRay = getSampleSize(sampleData.tileCoords.w) / abs(gradient);
+    float fixedStep = minComponent(sampleSizeAlongRay) * u_stepSize;
+
+    vec3 voxelCoord = sampleData.tileUv * vec3(u_dimensions);
+    vec3 directions = sign(gradient);
+    vec3 positiveDirections = max(directions, 0.0);
+    vec3 entryCoord = mix(ceil(voxelCoord), floor(voxelCoord), positiveDirections);
+    vec3 exitCoord = mix(floor(voxelCoord), ceil(voxelCoord), positiveDirections);
+
+    vec3 distanceFromEntry = abs(voxelCoord - entryCoord) * sampleSizeAlongRay;
+    vec3 distanceToExit = abs(exitCoord - voxelCoord) * sampleSizeAlongRay;
+
+    float lastEntry = minComponent(distanceFromEntry);
+    bvec3 isLastEntry = equal(distanceFromEntry, vec3(lastEntry));
+    vec3 normalShapeSpace = -1.0 * vec3(isLastEntry) * directions;
+    vec3 normal = normalize(jacobianT * normalShapeSpace);
+    vec4 voxelEntry = vec4(normal, currentT - lastEntry);
+    vec4 entry = intersectionMax(shapeIntersection.entry, voxelEntry);
+
+    float firstExit = minComponent(distanceToExit);
+    float stepSize = clamp(firstExit, fixedStep * 0.02, fixedStep);
```

Comment:
Where does the `0.02` come from? Should this be configurable, or at least declared as a constant?